controllers/awscluster_controller.go

```diff
@@ -477,16 +499,109 @@ func (r *AWSClusterReconciler) reconcileNormal(ctx context.Context, logger logr.
 		// will do that.
 	}
 
+	//
+	// Reconcile VPC endpoints
+	//
+	if !capiconditions.IsTrue(awsCluster, capa.ClusterSecurityGroupsReadyCondition) {
+		// Security groups are still not ready, we wait for them first
+		capiconditions.MarkFalse(awsCluster, VpcEndpointReady, ClusterSecurityGroupsNotReady, capi.ConditionSeverityWarning, "Security groups are still not ready")
+		return ctrl.Result{RequeueAfter: 5 * time.Minute}, nil
+	} else {
+		reconcileRequest := aws.ReconcileRequest[vpcendpoint.Spec]{
+			Resource:    awsCluster,
+			ClusterName: awsCluster.Name,
+			CloudResourceRequest: aws.CloudResourceRequest[vpcendpoint.Spec]{
+				RoleARN: roleArn,
+				Region:  awsCluster.Spec.Region,
+				Spec: vpcendpoint.Spec{
+					VpcId: awsCluster.Spec.NetworkSpec.VPC.ID,
+				},
+				AdditionalTags: awsCluster.Spec.AdditionalTags,
+			},
+		}
+		for _, subnet := range awsCluster.Spec.NetworkSpec.Subnets {
+			reconcileRequest.Spec.SubnetIds = append(reconcileRequest.Spec.SubnetIds, subnet.ID)
+		}
+		for _, securityGroup := range awsCluster.Status.Network.SecurityGroups {
```

Comment:
I actually saw a bug where it added only the default security group to the VPC endpoint, Might be a race condition issue

I think we should double check that there is expected amount of security group here and not only 1

Reply:
we need `mario0003-node` and `mario0003-controlplane` security group for sure, so maybe check that these two are present? if not wait for another reconciliation (1 min should be ok)

Reply:
The current logic is to wait that CAPA sets `ClusterSecurityGroupsReady` condition to true. Maybe that is not enough. I will look into what you suggested above and check again how CAPA sets security groups and how/when to be sure that everything we need is set there so we can use it.

Reply:
These 4:
```
var defaultAWSSecurityGroupRoles = []infrav1.SecurityGroupRole{
	infrav1.SecurityGroupAPIServerLB,
	infrav1.SecurityGroupLB,
	infrav1.SecurityGroupControlPlane,
	infrav1.SecurityGroupNode,
}
```

Plus optionally bastion SG.

Reply:
yep

Reply:
One thing that occurred to me, security group IDs are taken from AWSCluster.Status, while everything else is read from the AWSCluster.Spec. So maybe Status gets updated few times, and we just pick up some initial state (we are not updating VPC endpoint after initial creation, that is one of the TODOs/missing things, so implementing updates on reconciliation could make it work)

Reply:
hit the same issue again, so its not very rare and needs to be adresset